### PR TITLE
[clang][test] add testing for the AST matcher reference

### LINKED ISSUE #57607: classTemplateSpecializationDecl does not work as described
Hi,

I am trying to write a clang-tidy check and I have done this:
```cpp
void PotatoCheck::registerMatchers(MatchFinder *Finder) {
  Finder->addMatcher(classTemplateSpecializationDecl().bind("x"), this);
}

void PotatoCheck::check(const MatchFinder::MatchResult &Result) {
  const auto *MatchedDecl = Result.Nodes.getNodeAs<Decl>("x");
  diag(MatchedDecl->getLocation(), "instantiation");
}
```

The documentation for `classTemplateSpecializationDecl` says:
```txt
Given
  template<typename T> class A {};
  template<> class A<double> {};
  A<int> a;
classTemplateSpecializationDecl()
  matches the specializations A<int> and A<double>
```

But it doesn't. It matches the first two, not the last two:
```console
foo.cpp:1:30: warning: instantiation [misc-Potato]
  template<typename T> class A {};
                             ^
foo.cpp:2:20: warning: instantiation [misc-Potato]
  template<> class A<double> {};
                   ^
```

(foo.cpp contains precisely the three lines copied from the documentation.)

I am assuming I did something obviously wrong with the way I have set up the matcher? In which case I apologize for the noise.

### PR DESCRIPTION
**Pull request by 5chmidti**
## Problem Statement
Previously, the examples in the AST matcher reference, which gets generated by the doxygen comments in `ASTMatchers.h`, were untested and best effort.
Some of the matchers had no or wrong examples of how to use the matcher.

## Solution
This patch introduces a simple DSL around doxygen commands to enable testing the AST matcher documentation in a way that should be relatively easy to use.
In `ASTMatchers.h`, most matchers are documented with a doxygen comment. Most of these also have a code example that aims to show what the matcher will match, given a matcher somewhere in the documentation text. The way that the documentation is tested, is by using doxygen's alias feature to declare custom aliases. These aliases forward to `<tt>text</tt>` (which is what doxygen's `\c` does, but for multiple words). Using the doxygen aliases is the obvious choice, because there are (now) four consumers:
 - people reading the header/using signature help
 - the doxygen generated documentation
 - the generated html AST matcher reference
 - (new) the generated matcher tests

This patch rewrites/extends the documentation such that all matchers have a documented example.
The new `generate_ast_matcher_doc_tests.py` script will warn on any undocumented matchers (but not on matchers without a doxygen comment) and provides diagnostics and statistics about the matchers.

The current statistics emitted by the parser are:

```text
Statistics:
        doxygen_blocks                :   519
        missing_tests                 :    10
        skipped_objc                  :    42
        code_snippets                 :   503
        matches                       :   820
        matchers                      :   580
        tested_matchers               :   574
        none_type_matchers            :     6
```

The tests are generated during building and the script will only print something if it found an issue (compile failure, parsing issues, the expected and actual number of failures differs).

## Description

DSL for generating the tests from documentation.

TLDR:
```
  \header{a.h}
  \endheader     <- zero or more header

  \code
    int a = 42;
  \endcode
  \compile_args{-std=c++,c23-or-later} <- optional, the std flag supports std ranges and
                                          whole languages

  \matcher{expr()} <- one or more matchers in succession
  \match{42}   <- one or more matches in succession

  \matcher{varDecl()} <- new matcher resets the context, the above
                         \match will not count for this new
                         matcher(-group)
  \match{int a  = 42} <- only applies to the previous matcher (not to the
                         previous case)
```

The above block can be repeated inside a doxygen command for multiple code examples for a single matcher.
The test generation script will only look for these annotations and ignore anything else like `\c` or the sentences where these annotations are embedded into: `The matcher \matcher{expr()} matches the number \match{42}.`.

### Language Grammar
  [] denotes an optional, and <> denotes user-input

```
  compile_args j:= \compile_args{[<compile_arg>;]<compile_arg>}
  matcher_tag_key ::= type
  match_tag_key ::= type || std || count || sub
  matcher_tags ::= [matcher_tag_key=<value>;]matcher_tag_key=<value>
  match_tags ::= [match_tag_key=<value>;]match_tag_key=<value>
  matcher ::= \matcher{[matcher_tags$]<matcher>}
  matchers ::= [matcher] matcher
  match ::= \match{[match_tags$]<match>}
  matches ::= [match] match
  case ::= matchers matches
  cases ::= [case] case
  header-block ::= \header{<name>} <code> \endheader
  code-block ::= \code <code> \endcode
  testcase ::= code-block [compile_args] cases
```

### Language Standard Versions

The 'std' tag and '\compile_args' support specifying a specific language version, a whole language and all of its versions, and thresholds (implies ranges). Multiple arguments are passed with a ',' separator. For a language and version to execute a tested matcher, it has to match the specified '\compile_args' for the code, and the 'std' tag for the matcher. Predicates for the 'std' compiler flag are used with disjunction between languages (e.g. 'c || c++') and conjunction for all predicates specific to each language (e.g. 'c++11-or-later && c++23-or-earlier').

Examples:
 - `c`                                    all available versions of C
 - `c++11`                                only C++11
 - `c++11-or-later`                       C++11 or later
 - `c++11-or-earlier`                     C++11 or earlier
 - `c++11-or-later,c++23-or-earlier,c`    all of C and C++ between 11 and
                                          23 (inclusive)
 - `c++11-23,c`                             same as above

#### `type`:
**Match types** are used to select where the string that is used to check if a node matches comes from.
Available: `code`, `name`, `typestr`, `typeofstr`. The default is `code`.

 - `code`: Forwards to `tooling::fixit::getText(...)` and should be the preferred way to show what matches.
 - `name`: Casts the match to a `NamedDecl` and returns the result of `getNameAsString`. Useful when the matched AST node is not easy to spell out (`code` type), e.g., namespaces or classes with many members.
 - `typestr`: Returns the result of `QualType::getAsString` for the type derived from `Type` (otherwise, if it is derived from `Decl`, recurses with `Node->getTypeForDecl()`)

**Matcher types** are used to mark matchers as sub-matcher with 'sub' or as deactivated using 'none'. Testing sub-matcher is not implemented.

#### `count`:
Specifying a 'count=n' on a match will result in a test that requires that the specified match will be matched n times. Default is 1.

#### `std`:
A match allows specifying if it matches only in specific language versions. This may be needed when the AST differs between language versions.

#### `sub`:
The `sub` tag on a `\match` will indicate that the match is for a node of a bound sub-matcher.
E.g., `\matcher{expr(expr().bind("inner"))}` has a sub-matcher that binds to `inner`, which is the value for the `sub` tag of the expected match for the sub-matcher `\match{sub=inner$...}`. Currently, sub-matchers are not tested in any way.

#### ... I want to add a matcher?

Add a doxygen comment to the matcher with a code example, corresponding matchers and matches, that shows what the matcher is supposed to do. Specify the compile arguments/supported languages if required, and run `ninja check-clang-unit` to test the documentation.

#### ... the example I wrote is wrong?

The test-generation script will try to compile your example code before it continues. This makes finding issues with your example code easier because the test-failures are much more verbose.

The test-failure output of the generated test file will provide information about
 - where the generated test file is located
 - which line in `ASTMatcher.h` the example is from
 - which matches were: found, not-(yet)-found, expected
 - in case of an unexpected match: what the node looks like using the different `type`s
 - the language version and if the test ran with a windows `-target` flag (also in failure summary)

#### ... I don't adhere to the required order of the syntax?

The script will diagnose any found issues, such as `matcher is missing an example` with a `file:line:` prefix,
which should provide enough information about the issue.

#### ... the script diagnoses a false-positive issue with a doxygen comment?

It hopefully shouldn't, but if you, e.g., added some non-matcher code and documented it with doxygen, then the script will consider that as a matcher documentation. As a result, the script will print that it detected a mismatch between the actual and the expected number of failures. If the diagnostic truly is a false-positive, change the `expected_failure_statistics` at the top of the `generate_ast_matcher_doc_tests.py` file.

Fixes #57607
Fixes #63748

### CODE REVIEW

**Comment by 5chmidti** [comment]
CC @llvm/pr-subscribers-clang-tidy as stake-holders in matchers

**Review by AaronBallman — COMMENTED** [review]
It looks like precommit CI found relevant failures:
```
_bk;t=1717501095887FAILED: tools/clang/unittests/ASTMatchers/ASTMatchersDocTests.cpp 
_bk;t=1717501095887cmd.exe /C "cd /D C:\ws\src\build\tools\clang\unittests\ASTMatchers && C:\ws\src\clang\utils\generate_ast_matcher_doc_tests.py --input-file C:/ws/src/clang/include/clang/ASTMatchers/ASTMatchers.h --output-file C:/ws/src/build/tools/clang/unittests/ASTMatchers/ASTMatchersDocTests.cpp"
_bk;t=1717501095887  File "C:\ws\src\clang\utils\generate_ast_matcher_doc_tests.py", line 613
_bk;t=1717501099131    const StringRef Code = R"cpp(\n{"\t#include \"cuda.h\"\n" if has_cuda else ""}{self.code})cpp";\n"""

_bk;t=1717501099131                                                                                                        ^

_bk;t=1717501099131SyntaxError: f-string expression part cannot include a backslash
```

**Review by AaronBallman — COMMENTED** [review]
I wonder if we should add some documentation to ASTMatchers.h about the new special syntax for comments so that users who hit test failures with the new automatic tests have some more help getting to a solution.

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 1]
```diff
@@ -330,35 +377,46 @@ AST_POLYMORPHIC_MATCHER_P(isExpandedFromMacro,
...
 ///   void X();
 ///   class C {
-///     friend X;
+///     friend void X();
 ///   };
 /// \endcode
+/// \compile_args{-std=c++}
+/// The matcher \matcher{decl()}
+/// matches \match{void X()}, \match{type=name;count=2$C}
+/// and \match{count=2$friend void X()}.
```
[the remark this answers is not in the document]
Comment (5chmidti) [reply clang/include/clang/ASTMatchers/ASTMatchers.h:390]:
```
|-FunctionDecl <line:1:1, col:8> col:6 X 'void ()'
`-CXXRecordDecl <line:2:1, line:4:1> line:2:7 class C definition
  |-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
  | |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
  | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveConstructor exists simple trivial needs_implicit
  | |-CopyAssignment simple trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveAssignment exists simple trivial needs_implicit
  | `-Destructor simple irrelevant trivial needs_implicit
  |-CXXRecordDecl <col:1, col:7> col:7 implicit class C
  `-FriendDecl <line:3:5, col:19> col:17
    `-FunctionDecl parent 0xf23a388 prev 0xf284370 <col:5, col:19> col:17 friend X 'void ()'
```

> Can you explain `\match{type=name;count=2$C}`?

That is the `implicit class C` in the AST above. I couldn't access it from the top-level `C` and I couldn't find a way from the `implicit class C` back to the top-level one, so I don't know how to call it. I thought it would be a decl but not a definition, however, `getDefinition` returns a `nullptr` for the `implicit class C`.

> should we add a comment explaining that other match?

Certainly. I'll read the documentation again to see if there are more cases like this that could be improved as well.

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 2]
```diff
@@ -176,11 +176,13 @@ inline internal::TrueMatcher anything() { return internal::TrueMatcher(); }
...
 ///   namespace NS {
-///   int Y;
+///     int Y;
 ///   }  // namespace NS
 /// \endcode
-/// decl(hasDeclContext(translationUnitDecl()))
-///   matches "int X", but not "int Y".
+/// \compile_args{-std=c++}
+/// The matcher \matcher{namedDecl(hasDeclContext(translationUnitDecl()))}
+/// matches \match{type=name$X} and \match{type=name$NS},
```
[the remark this answers is not in the document]
Comment (5chmidti) [reply clang/include/clang/ASTMatchers/ASTMatchers.h:184]: Using `type=name` can be generally considered to be a style/readability/expressiveness choice if the AST node supports it. The `X` example would probably be better spelling the declaration out, the same goes for `Y` (probably remnants of the early days). There may be other trivial examples that could be spelled out. 

There are for sure some more trivial cases which could be spelled out. I'll check on the documentation again tomorrow and provide some updates (also w.r.t to your other comment).

If we wanted to spell out the namespace, we could, but that would require writing the `NS` in a single line. It's an artificial limitation in the script that can probably be implemented if we want to have the option.

**Review by 5chmidti — COMMENTED** [review]
> I wonder if we should add some documentation to ASTMatchers.h

I'll add one in a day or so. I updated the description with some more information, and I'll probably take parts of that as a basis for the comment in the header (and update the script comment as well).

> so that users who hit test failures with the new automatic tests have some more help getting to a solution.

There is now a `What if ...?` section to the pr description, which I will put into the header comment as well.

File: clang/include/clang/ASTMatchers/ASTMatchers.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/include/clang/ASTMatchers/ASTMatchers.h:390]: okay, thank you! I kind of figured it was the implicit class declaration.

**Comment by 5chmidti** [comment]
- added a file-level comment in the ASTMatcher.h file on how the syntax works (basically the pr description)
- replaced some `type=name` matches with explicit code matches where applicable, to be more expressive
- added comments to `count=` matches when they didn't explain why or if there were multiple matches

**Review by AaronBallman — COMMENTED** [review]
The changes generally LGTM, though I would appreciate a second set of eyes on the CMake and Python changes because I have a bit less confidence in my review abilities there.

Thank you for adding the documentation to the header file, I think that will help folks when working on their own matchers.

One question I have is: do you happen to know how this impacts build times for Clang itself? I'm assuming that if ASTMatchers.h isn't modified, CMake won't re-run `generate_ast_matcher_doc_tests.py` and so the compile time performance hit is only on full rebuilds or when changing the header?

**Comment by 5chmidti** [comment]
Thanks.

> ... how this impacts build times for Clang itself? I'm assuming that if ASTMatchers.h isn't modified, CMake won't re-run generate_ast_matcher_doc_tests.py and so the compile time performance hit is only on full rebuilds or when changing the header?

The 'state' of the generated file is only checked when the `ASTMatchersTests` target is built, because it's the only thing that depends on the generated file. And the file is only generated when: the file does not exist or `ASTMatchers.h` has changed (excluding transitive changes in includes) or `generate_ast_matcher_doc_tests.py` has changed.

**Comment by AaronBallman** [comment]
> Thanks.
> 
> > ... how this impacts build times for Clang itself? I'm assuming that if ASTMatchers.h isn't modified, CMake won't re-run generate_ast_matcher_doc_tests.py and so the compile time performance hit is only on full rebuilds or when changing the header?
> 
> The 'state' of the generated file is only checked when the `ASTMatchersTests` target is built, because it's the only thing that depends on the generated file. And the file is only generated when: the file does not exist or `ASTMatchers.h` has changed (excluding transitive changes in includes) or `generate_ast_matcher_doc_tests.py` has changed.

Excellent, thank you for the confirmation! That sounds reasonable to me.

**Comment by 5chmidti** [comment]
> Do you need me to land the changes on your behalf?

No need, I can do that. However, this PR is in a stack and depends on two small PRs: #94244 and #94243 that need to be reviewed before this PR can be merged

**Comment by 5chmidti** [comment]
rebase on trunk + rebased stack
